apps/grades-sync/src/job.rs

```diff
@@ -0,0 +1,203 @@
+        Ok(())
+    }
+
+    async fn perform_subject_synchronization(&self) -> anyhow::Result<()> {
+        info!("performing subject synchronization");
+        let subjects = get_subjects().await?;
+        let subject_count = subjects.len();
+        info!(
+            "performing synchronization for {} subjects",
+            subject_count
+        );
+
+        futures::future::join_all(subjects
+            .into_iter()
+            .map(|subject| {
+                async move {
+                    self.synchronize_single_subject(subject).await.unwrap();
+                }
+            })).await;
+        info!("synchronization complete");
+        Ok(())
+    }
+
+    async fn perform_grade_synchronization(&self) -> anyhow::Result<()> {
+        info!("performing grade synchronization");
+        let grades = get_grades().await?;
+        info!(
+            "performing synchronization for {} grades",
+            grades.len()
+        );
+        let map = grades.into_iter()
+            .filter(|grade| grade
+                .total_candidates
+                .parse::<i32>()
+                .expect("invalid number for total candidates") > 0
+            )
+            .into_grouping_map_by(|grade| grade.subject_code.clone())
+            // TODO: Maybe there is a way to avoid this collect?
+            .collect::<Vec<_>>();
+        futures::future::join_all(map.into_values().map(|grades| {
+                async move {
+                    for grade in grades {
+                        self.synchronize_single_grade(grade).await.unwrap();
+                    }
+                }
+            })).await;
+        info!("synchronization complete");
+        Ok(())
+    }
+
+    async fn synchronize_single_faculty(&self, department: HkdirDepartment) -> anyhow::Result<()> {
+        // Create the faculty if it doesn't exist. The underlying query performs an on
+        // conflict update, which means that we do not need to check if the faculty
+        // exists before creating it.
+        let faculty = self
+            .faculty_repository
+            .create_faculty(department.faculty_code, department.faculty_name)
+            .await
+            .unwrap();
+
+        // Create the department. The same rules apply here as for the faculty.
+        self.department_repository
+            .create_department(
+                department.department_name,
+                department.department_code,
+                faculty.id,
+            )
+            .await
+            .unwrap();
+        Ok(())
+    }
+
+    async fn synchronize_single_subject(&self, subject: HkdirSubject) -> anyhow::Result<()> {
+        // Forge a slug from the subject name. A lot of the subject code fields from
+        // HKDir have a -1 or another number appended to them, which is not useful
+        // because we're using to seeing 'TDT4120' instead of 'TDT4120-1'.
+        let re = Regex::new("-[A-Za-z0-9]+$").unwrap();
```

Comment:
this compiles the regex every time this function is called, not sure how much time it takes but could be worth it to move it into a global using `lazy_static` or something

Reply:
Yep makes sense